### ISSUE TITLE
`redirectTo` not being respected when using email or magic link auth

### ISSUE THREAD

**Issue by johnhforrest** [supabase/supabase#4540]
# Bug report

## Describe the bug

I've been stuck on this for hours and thought to finally reach out. It looks like there are a bunch of old bug reports that indicate this should be working now but I haven't been able to get it to work.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:
```
const { error } = await supabase.auth.signIn(
      { email, password },
      { redirectTo: 'http://localhost:3000/dashboard' },
);
```

I have http://localhost:3000/dashboard configured as an Additional Redirect URL on my Supabase project settings. I see the `redirect_to` parameter in the /auth/v1/token request to Supabase but I always get bounced back to http://localhost:3000 and not the dashboard page.

## Expected behavior
`redirectTo` would go to the expected page.

## Screenshots

Not sure how helpful screenshots would be, but I can see that the referrer to the request back to index is http://localhost:3000/signin. I've tried putting breakpoints on the GoTrue client code but the decision on where to redirect appears to happen on the server.

## System information

- OS: mac
- Browser (if applies) chrome
- Version of supabase-js: `"@supabase/supabase-js": "^1.28.5",`
- Version of Node.js: 16.2.0
- Version of Next.js: `"next": "^12.0.7",`

## Additional context

In general the Supabase auth docs could use some love. They are really high level but don't provide a lot of details so I ended up just following a lot of the example code and things seem iffy at best. Maybe it's the Next.js integration? My users get logged out all of the time--perhaps I need to extend the JWT expiration? Having some examples showing persistent session would be appreciated or examples of how supabase.io handles auth so I don't have to enter my credentials every day.

**Comment by rmrt1n** [comment]
I'm having the same issue. My workaround for this is to add the redirect logic in the onAuthStateChange function.
```js
supabase.auth.onAuthStateChange(async (event, session) => {
  if (event !=== 'SIGNED_OUT') {
    Router.push('/profile');
  } else {
    Router.push('/signin');
  }
});
```
Something like this. It works for now but it'll be better if this is handled by `redirectTo` directly.

**Comment by TheThirdRace** [comment]
I would love for the Magic Link to respect the `redirectTo`.

When a user is in a page accessible from a long/deep route, ex: `XXXX/XX/XXXX/XX/XXXX/XX/XXXX/XX/`, it's no fun at all to be redirected back to the home page...

All the providers are redirecting to the correct page, only Magic Links are stuck on the home page only.

**Comment by tatwater** [comment]
I've had this problem with OAuth as well! Only tested with the google provider, but exact same behavior as described above.

**Comment by J0** [comment]
Hey all,

At present it seems like the ` redirectTo` param is only respected for OAuth signIn: [as per the docs](https://supabase.com/docs/reference/javascript/auth-signin#sign-in-with-redirect)

Think that this could be clearer on the docs and I'll raise it during the next Auth meeting -- apologies for any confusion is caused.

With regards to OAuth, I've validated that this is an issue for the Github provider and will raise it during the next Supabase Auth meeting.

Appreciate your patience thus far, let us know if you have any further questions or concerns.

Thanks!
Joel

**Comment by TheThirdRace** [comment]
> At present it seems like the ` redirectTo` param is only respected for OAuth signIn: [as per the docs](https://supabase.com/docs/reference/javascript/auth-signin#sign-in-with-redirect)

@J0 
But the question is, can it be fixed?

Because while the doc says we should send the correct link in the email, as far as I'm aware, we don't have that information in the magic link email Supabase sends by itself...

Meaning, given the current state of Magic Links, we can't modify the email template to include the `redirectTo` we need.

Are you suggesting we should simply give up on that and build our own Magic Links? 

Sorry, your answer was a bit ambiguous, I'm just trying to figure out what it actually mean.

**Comment by lack-of-gravitas** [comment]
Are there any plans / ETA on this specific issue? Currently affecting clients that only want Magic Link as Auth method.

**Comment by J0** [comment]
Hey @TheThirdRace @lack-of-gravitas,

Apologies for the delayed reply and perhaps unclear response. We do have plans to better support for redirects with magic link by having clearer types so keep your eyes peeled. 

At present, redirects with magic link of the form:

```js
async function signInWithEmail() {
  const { user, session, error } = await supabase.auth.signIn({
    email: "joel@supabase.io"
  }, {
      redirectTo: "https://www.google.com"
    })
}
```
Should be supported. If you're still facing issues with this please feel free to shoot over an email to support@supabase.io

**Comment by lack-of-gravitas** [comment]
Thanks @J0 -- that's what I'm rolling with but doesnt work at the moment in a basic implementation but the confirmation link I receive is still the base path. This segment / entire code is from the supabase/examples/nextjs-subscription-payments repo.

See redirect path  in the Magic Link email -- that's the base path. It should be basePath + /courses/couch-to-5km for the bit I'm testing. Also have included Site and Redirect URLs in Supabase Auth settings and the email template in case there's some magic that can be applied there.

![image](https://user-images.githubusercontent.com/10094964/184530941-a3608208-312c-4a02-9ad6-864ac2b052b1.png)

![image](https://user-images.githubusercontent.com/10094964/184531003-5b62dce1-3a99-40e7-86c7-bc91ccdfd0fd.png)
![image](https://user-images.githubusercontent.com/10094964/184531063-43145174-439f-4f8f-86b0-91fee0e0dd3b.png)

![image](https://user-images.githubusercontent.com/10094964/184530657-2d6e2b82-c978-4c41-a2c4-8cb25c51d385.png)

**Comment by TheThirdRace** [comment]
@J0 
> Could you please confirm that the URL you wish to redirect to is under the configured list of valid redirect URLs?

Just as @lack-of-gravitas  wrote, I also configured redirect URLs with wildcards.

The feature we need is to be redirected to the same page you were before sending the magic link.

If my website has 10,000 pages, I need to be able to redirect to any of those 10,000 pages without listing them all one by one in the allowed list...

At the moment, no matter which page I'm on, the magic link always redirect to the home page even if I pass the right URL to the `redirectTo` parameter.

**Comment by J0** [comment]
@TheThirdRace @lack-of-gravitas 

Sorry to hear that the redirects aren't working as expected -- it can get frustrating fumbling around with the config. Quick check that I'm understanding both of you correctly, would this be the expected behaviour? 

1. App sends an email from a given URL, let's say as a contrived example: `https://klug.vercel.app/courses/course-to-5km/`
2. We call

```js
async function signInWithEmail() {
  const { user, session, error } = await supabase.auth.signIn({
    email: "awesomeuser@gmail.com"
  }, {
      redirectTo: "https://klug.vercel.app/courses/course-to-5km/"
    })
}
```

4. User receives email and clicks on the link in the email
5. User should now be signed in and redirected to `https://klug.vercel.app/courses/course-to-5km/`

This should be possible with Supabase Auth though the globbing might need some adjustments(it can be confusing, and we're doing our best to make it more digestible). Could you please try adding `https://klug.vercel.app/**` as a redirect URL and then trying again?

**Comment by TheThirdRace** [comment]
@J0 

Your description of what is expected is accurate. This is exactly what we want.

I just tested with your solution using `base URL` + `/**`  and it worked.

It would be a good idea to include this pattern detail in the settings page so people are aware of it.

Thank you very much for your help, you've made my day

**Comment by J0** [comment]
@TheThirdRace 

Thank you for your patience with me.

Glad to hear that the issue has been resolved -- we'll be sure to highlight this to the docs team so that it can be captured in the settings or the docs.

Have a great day ahead as well!

**Comment by lack-of-gravitas** [comment]
Yes @J0 this is the behaviour we expect -- I did exactly that and that works now too.
Thank you again for the quick around... much love to Supabase Team.

Might be overkill but here's what I've got as redirects in Supabase Auth configuration screen for any future people struggling with this.

![image](https://user-images.githubusercontent.com/10094964/184578357-84cf9f73-cf92-4cb8-abf0-c476feacd758.png)
